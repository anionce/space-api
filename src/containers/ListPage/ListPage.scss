@import '../../assets/styles/variables';

.list-container {
	margin-top: 32px;
	display: flex;
	align-items: center;
	flex-direction: column;
}
.loading {
	margin: 32px;
	text-align: center;
	font-weight: 800;
	text-decoration: underline $color-vidid;
	text-decoration-thickness: 5px;
	font-size: larger;
}
