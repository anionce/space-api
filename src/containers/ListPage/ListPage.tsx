import { Card } from '../../components/Card';
import { Pagination } from '../../components/Pagination';
import { Input } from '../../components/Input';
import { TypeFilter } from '../../components/TypeFilter';
import { resultMatch } from '../../helpers/helpers';
import { useFilter } from '../../hooks/useFilter';
import { usePagination } from '../../hooks/usePagination';
import { useSearch } from '../../hooks/useSearch';
import { useListCapsulesQuery } from '../../services/api';
import './ListPage.scss';
import { ErrorPage } from '../ErrorPage';
import { useAppSelector } from '../../store/hooks';
import { capsuleFilterTypes } from '../../models/Capsule';
import { Item } from '../../models/common';
import { rocketFilterTypes } from '../../models/Rocket';
import { Capsule } from '../../models/Capsule';
import { Rocket } from '../../models/Rocket';
import { useLocation } from 'react-router-dom';
import React from 'react';
import { NotFound } from '../NotFound/NotFound';
import { RootState } from 'src/store/store';

export const ListPage = () => {
	const { pathname } = useLocation();

	const { data: capsules, isFetching, isError } = useListCapsulesQuery();
	const rockets: Rocket[] = useAppSelector((state: RootState) => state.rocketSlice.rockets);
	const isCapsule: Boolean = pathname.includes('capsules');

	const data: Rocket[] | Capsule[] | undefined = isCapsule ? capsules : rockets;

	const { onInputChange, searchedItems, searchText, emptySearch } = useSearch(data);
	const { filteredItems, emptyFilter, onTypeChanged } = useFilter(data);

	const searchedArray: Item[] | undefined = !emptySearch ? searchedItems : data;
	const filteredArray: Item[] | undefined = !emptyFilter ? filteredItems : data;
	const finalResult: Item[] = resultMatch(searchedArray, filteredArray);

	let { currentItems, paginate, currentPage, pagesPerPage } = usePagination(finalResult);

	return (
		<>
			<Input searchText={searchText} onInputChange={onInputChange} />
			<TypeFilter
				onTypeChanged={onTypeChanged}
				itemFilterTypes={isCapsule ? capsuleFilterTypes : rocketFilterTypes}
				defaultOption={'All'}
			/>
			{isFetching && <div className='loading'>Loading...</div>}
			{isError && <ErrorPage />}
			{currentItems && currentItems?.length === 0 && <NotFound />}
			<div className='list-container'>
				{currentItems?.map((item: Rocket) => (
					<Card key={item.id} item={item} />
				))}
				<Pagination
					count={finalResult?.length}
					paginate={paginate}
					pagesPerPage={pagesPerPage}
					currentPage={currentPage}
				/>
			</div>
		</>
	);
};
