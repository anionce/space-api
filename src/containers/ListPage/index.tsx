import { ListPage } from './ListPage';
export { ListPage };
