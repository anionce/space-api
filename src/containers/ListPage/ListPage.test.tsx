import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { screen } from '@testing-library/react';
import { customRender } from '../../assets/tests/utils';
import { ListPage } from './ListPage';

test('renders correcly', () => {
	const { container } = customRender(<ListPage />);

	expect(container).toMatchSnapshot();
});

test('all elements are displayed', () => {
	customRender(<ListPage />);

	expect(screen.getAllByText(/Details/i)).toHaveLength(5);
});
