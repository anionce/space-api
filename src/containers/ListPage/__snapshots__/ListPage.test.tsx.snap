// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correcly 1`] = `
<div>
  <div
    class="input-container"
  >
    <label
      for="input"
    >
      Search
    </label>
    <input
      id="input"
      maxlength="100"
      value=""
    />
  </div>
  <div
    class="select-container"
  >
    <label
      for="type"
    >
      Type
    </label>
    <select
      id="type"
    >
      <option
        value="All"
      >
        All
      </option>
      <option
        value="SuperRocket"
      >
        SuperRocket
      </option>
      <option
        value="HiperRocket"
      >
        HiperRocket
      </option>
      <option
        value="ExtraRocket"
      >
        ExtraRocket
      </option>
    </select>
  </div>
  <div
    class="loading"
  >
    Loading...
  </div>
  <div
    class="list-container"
  >
    <div
      class="card"
    >
      <div
        class="row"
      >
        <div
          class="title-container"
        >
          <div>
            Rocky 1
          </div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="RocketIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2.5s4.5 2.04 4.5 10.5c0 2.49-1.04 5.57-1.6 7H9.1c-.56-1.43-1.6-4.51-1.6-7C7.5 4.54 12 2.5 12 2.5zm2 8.5c0-1.1-.9-2-2-2s-2 .9-2 2 .9 2 2 2 2-.9 2-2zm-6.31 9.52c-.48-1.23-1.52-4.17-1.67-6.87l-1.13.75c-.56.38-.89 1-.89 1.67V22l3.69-1.48zM20 22v-5.93c0-.67-.33-1.29-.89-1.66l-1.13-.75c-.15 2.69-1.2 5.64-1.67 6.87L20 22z"
            />
          </svg>
        </div>
        <div
          class="subtitle"
        >
          SuperRocket
        </div>
      </div>
      <div
        class="row"
      >
        <a
          href="/rockets/1"
        >
          <button
            name="detailsButton"
          >
            Details
          </button>
        </a>
      </div>
    </div>
    <div
      class="card"
    >
      <div
        class="row"
      >
        <div
          class="title-container"
        >
          <div>
            Rocket 2
          </div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="RocketIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2.5s4.5 2.04 4.5 10.5c0 2.49-1.04 5.57-1.6 7H9.1c-.56-1.43-1.6-4.51-1.6-7C7.5 4.54 12 2.5 12 2.5zm2 8.5c0-1.1-.9-2-2-2s-2 .9-2 2 .9 2 2 2 2-.9 2-2zm-6.31 9.52c-.48-1.23-1.52-4.17-1.67-6.87l-1.13.75c-.56.38-.89 1-.89 1.67V22l3.69-1.48zM20 22v-5.93c0-.67-.33-1.29-.89-1.66l-1.13-.75c-.15 2.69-1.2 5.64-1.67 6.87L20 22z"
            />
          </svg>
        </div>
        <div
          class="subtitle"
        >
          ExtraRocket
        </div>
      </div>
      <div
        class="row"
      >
        <a
          href="/rockets/2"
        >
          <button
            name="detailsButton"
          >
            Details
          </button>
        </a>
      </div>
    </div>
    <div
      class="card"
    >
      <div
        class="row"
      >
        <div
          class="title-container"
        >
          <div>
            Rocket 3
          </div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="RocketIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2.5s4.5 2.04 4.5 10.5c0 2.49-1.04 5.57-1.6 7H9.1c-.56-1.43-1.6-4.51-1.6-7C7.5 4.54 12 2.5 12 2.5zm2 8.5c0-1.1-.9-2-2-2s-2 .9-2 2 .9 2 2 2 2-.9 2-2zm-6.31 9.52c-.48-1.23-1.52-4.17-1.67-6.87l-1.13.75c-.56.38-.89 1-.89 1.67V22l3.69-1.48zM20 22v-5.93c0-.67-.33-1.29-.89-1.66l-1.13-.75c-.15 2.69-1.2 5.64-1.67 6.87L20 22z"
            />
          </svg>
        </div>
        <div
          class="subtitle"
        >
          SuperRocket
        </div>
      </div>
      <div
        class="row"
      >
        <a
          href="/rockets/3"
        >
          <button
            name="detailsButton"
          >
            Details
          </button>
        </a>
      </div>
    </div>
    <div
      class="card"
    >
      <div
        class="row"
      >
        <div
          class="title-container"
        >
          <div>
            Rocket 4
          </div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="RocketIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2.5s4.5 2.04 4.5 10.5c0 2.49-1.04 5.57-1.6 7H9.1c-.56-1.43-1.6-4.51-1.6-7C7.5 4.54 12 2.5 12 2.5zm2 8.5c0-1.1-.9-2-2-2s-2 .9-2 2 .9 2 2 2 2-.9 2-2zm-6.31 9.52c-.48-1.23-1.52-4.17-1.67-6.87l-1.13.75c-.56.38-.89 1-.89 1.67V22l3.69-1.48zM20 22v-5.93c0-.67-.33-1.29-.89-1.66l-1.13-.75c-.15 2.69-1.2 5.64-1.67 6.87L20 22z"
            />
          </svg>
        </div>
        <div
          class="subtitle"
        >
          HiperRocket
        </div>
      </div>
      <div
        class="row"
      >
        <a
          href="/rockets/4"
        >
          <button
            name="detailsButton"
          >
            Details
          </button>
        </a>
      </div>
    </div>
    <div
      class="card"
    >
      <div
        class="row"
      >
        <div
          class="title-container"
        >
          <div>
            Rocket 5
          </div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="RocketIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2.5s4.5 2.04 4.5 10.5c0 2.49-1.04 5.57-1.6 7H9.1c-.56-1.43-1.6-4.51-1.6-7C7.5 4.54 12 2.5 12 2.5zm2 8.5c0-1.1-.9-2-2-2s-2 .9-2 2 .9 2 2 2 2-.9 2-2zm-6.31 9.52c-.48-1.23-1.52-4.17-1.67-6.87l-1.13.75c-.56.38-.89 1-.89 1.67V22l3.69-1.48zM20 22v-5.93c0-.67-.33-1.29-.89-1.66l-1.13-.75c-.15 2.69-1.2 5.64-1.67 6.87L20 22z"
            />
          </svg>
        </div>
        <div
          class="subtitle"
        >
          SuperRocket
        </div>
      </div>
      <div
        class="row"
      >
        <a
          href="/rockets/5"
        >
          <button
            name="detailsButton"
          >
            Details
          </button>
        </a>
      </div>
    </div>
    <div
      class="pagination"
    >
      <div
        class="active"
        title="pageNumber"
      >
        1
      </div>
      <div
        class="page"
        title="pageNumber"
      >
        2
      </div>
    </div>
  </div>
</div>
`;
